src/ros2idl.ne

```diff
@@ -0,0 +1,395 @@
+const keywords = [
+  , "struct"
+  , "module"
+  , "const"
+  , "include"
+
+  //types
+  , "boolean"
+  , "wstring"
+  , "string"
+  , "sequence"
+	
+  // numeric types
+  , "byte"
+  , "octet"
+  , "wchar"
+  , "char"
+  , "double"
+  , "float"
+  , "int8"
+  , "uint8"
+  , "int16"
+  , "uint16"
+  , "int32"
+  , "uint32"
+  , "int64"
+  , "uint64"
+  , "unsigned" 
+  , "short"  
+  , "long"
+];
+
+const kwObject = keywords.reduce((obj, w) => {
+	obj[w] = w;
+	return obj;
+}, {});
+
+const moo = require("moo");
+// Terminal tokens are in all caps
+const lexer = moo.compile({
+  SPACE: {match: /\s+/, lineBreaks: true},
+  DECIMALEXP: /(?:(?:\d+\.\d*)|(?:\d*\.\d+)|(?:[0-9]+))[eE](?:[+|-])?[0-9]+/,
+  DECIMAL: /(?:(?:\d+\.\d*)|(?:\d*\.\d+))/,
+  INTEGER: /[0-9]+/,
+  COMMENT: /(?:\/\/[^\n]*)|(?:\/\*(?:.|\n)+?\*\/)/,
+  HEX_LITERAL: /0x(?:[0-9][a-f][A-F])+?/,
+  STRING: [{match: /"(?:\\["\\rnu]|[^"\\])*?"/, lineBreaks: true, value: x => x.slice(1, -1)}], // remove outside quotes
+  LCBR: '{',
+  RCBR: '}',
+  LBR: '[',
+  RBR: ']',
+  LT: '<',
+  GT: '>',
+  LPAR: '(',
+  RPAR: ')',
+  ';': ';',
+  ',': ',',
+  AT: '@',
+  PND: '#',
+  PT: ".",
+  '/': "/",
+  SIGN: /[+|-]/,
+  EQ: /=[^\n]*?/,
+  NAME: {match: /[a-zA-Z_][a-zA-Z0-9_]*(?:\:\:[a-zA-Z][a-zA-Z0-9_]*)*/, type: moo.keywords(kwObject)},
+});
+
+// Utiility functions
+
+// also used to parse tokens to strings since they start as an object
+function join(d){
+	return d.join("");
+}
+
+// used for combining AST components
+function extend(objs) {
+  return objs.reduce((r, p) => ({ ...r, ...p }), {});
+}
+
+function noop() {
+	return null;
+}
+
+// Constants can be used in defs to define lengths of arrays and strings
+// they should be defined before the def that uses them
+const constantToValueMap = {};
+
+function getIntOrConstantValue(d) {
+	const int = parseInt(d);
+	if(!isNaN(int)) {
+		return int
+	}
+
+	// handle %NAME value
+	return d?.value ? constantToValueMap[d.value] : undefined;	
+}
+
+function processComplexModule(d) {
+  const moduleName = d[0][4].name;
+  const defs = d[0][8];
+  // returning array of message definitions
+  const msgDefs = [];
+  function traverse(node, processNode) {
+    if(Array.isArray(node)) {
+      node.forEach(n => traverse(n, processNode));
+    } else {
+      processNode(node);
+    }
+  }
+  // Need to update the names of modules and structs to be in their respective namespaces
+  traverse(d[0], (sub) => {
+    if(sub && sub.definitions) {
+      sub.name = `${moduleName}::${sub.name}`;
+      msgDefs.push(sub);
+    }
+  });
+  
+  
+  return msgDefs;
+}
+
+function processConstantModule(d) {
+	const moduleName = d[0][4].name;
+  const enclosedConstants = d[0][8];
+  // need to return array here to keep same signature as processComplexModule
+	return [{
+		name: moduleName,
+		definitions: enclosedConstants.flat(1),
+	}];
+}
+
+%}
+
+@lexer lexer
+
+main -> (importDcl __):* module:+ _ {% d => {
```

Comment:
yeah I can update this so that they can be between modules.  though I'm not quite sure we should allow include statements in this context since the schemas in idl should be standalone right? We won't be able to import these so should we disallow them?

Reply:
With the format described [here](https://mcap.dev/specification/appendix.html#ros2idl), are we going to see concatenated files that contain includes?

Reply:
I think so given the [test case here: ](https://github.com/ros2/rosbag2/blob/8044dba28aa7f99247ccf29861f349969be07117/rosbag2_storage_mcap/test/rosbag2_storage_mcap/test_message_definition_cache.cpp#L63-L126), but they can be ignored by the parser because the definition within the include should be in the same file.

Reply:
That's what I'd expect yeah. So my question here is: can imports and module definitions ever be interleaved in a way that we'd need to change this to something like `(importDcl | module):+`?

Reply:
what I have should allow this now as long as a `module` follows the `#include` ie: modules can't be at the end of files. This makes sense because imported "building blocks" (modules, structes, etc), need to be declared before they are used accored to spec.